Kernel/setSubstitutionSystem$wl.m

```diff
@@ -348,42 +346,41 @@ All operations here should evaluate in O(1).*)
 	outputWithMetadata = Catch[
 		Reap[
 			setReplace$wl[setWithMetadata, rulesWithMetadata, stepSpec, vertexIndex, returnOnAbortQ, timeConstraint],
-			{$$deletedExpressions, $$ruleIndex}],
+			{$$deletedExpressions, $$events}],
 		$$nonListExpression,
 		(makeMessage[caller, "nonListExpressions", #];
-			Return[$Failed]) &]; (* {{finalState, terminationReason}, {deletedExpressions}} *)
+			Return[$Failed]) &]; (* {{finalState, terminationReason}, {{deletedExpressions}, {events}}} *)
 	If[outputWithMetadata[[1]] === $Aborted, Return[$Aborted]];
-	result = SortBy[
+	allExpressions = SortBy[
 		Join[
 			outputWithMetadata[[1, 1]],
 			If[outputWithMetadata[[2, 1]] == {}, {}, outputWithMetadata[[2, 1, 1]]]],
 		First];
-	intermediateEvolution = WolframModelEvolutionObject[<|
-		$creatorEvents -> result[[All, 2]],
-		$destroyerEvents -> result[[All, 3]],
-		$generations -> result[[All, 4]],
-		$atomLists -> result[[All, 5]],
+	initialEvent = {0, {} -> setWithMetadata[[All, 1]], 0};
+	allEvents = Join[{initialEvent}, If[outputWithMetadata[[2, 2]] == {}, {}, outputWithMetadata[[2, 2, 1]]]];
+	generationsCount = Max[allEvents[[All, 3]], 0];
+	maxCompleteGenerationAssumingExceedingStepSpecGenerationsDontExist = CheckAbort[
+		maxCompleteGeneration[outputWithMetadata[[1, 1]], renamedRules],
+		If[returnOnAbortQ,
+			Missing["Unknown", $Aborted],
+			Return[$Aborted]
+		]];
+
+	WolframModelEvolutionObject[<|
+		$version -> 2,
 		$rules -> rules,
-		$maxCompleteGeneration -> CheckAbort[
-			maxCompleteGeneration[outputWithMetadata[[1, 1]], renamedRules],
-			If[returnOnAbortQ,
-				Missing["Unknown", $Aborted],
-				Return[$Aborted]
-			]],
-		$terminationReason -> outputWithMetadata[[1, 2]],
-		$eventRuleIDs -> If[outputWithMetadata[[2, 2]] == {}, {}, outputWithMetadata[[2, 2, 1]]]|>];
-	WolframModelEvolutionObject[
-		Join[
-			intermediateEvolution[[1]],
-			<|$maxCompleteGeneration -> With[{
-					possibleInfinityResult = intermediateEvolution[[1, Key[$maxCompleteGeneration]]]},
-				If[MissingQ[possibleInfinityResult],
-					possibleInfinityResult,
-					Min[possibleInfinityResult, intermediateEvolution["TotalGenerationsCount"]]]],
-			$terminationReason -> Replace[
-				intermediateEvolution[[1, Key[$terminationReason]]],
-				$fixedPoint ->
-					If[intermediateEvolution["TotalGenerationsCount"] == Lookup[stepSpec, $maxGenerationsLocal, Infinity],
-						$maxGenerationsLocal,
-						$fixedPoint]]|>]]
+		$maxCompleteGeneration -> If[MissingQ[maxCompleteGenerationAssumingExceedingStepSpecGenerationsDontExist],
```

Comment:
Now _that_ is a variable name...

Reply:
It's too obscure. I tried to come up with a better name. Suggestions are welcome.

Also, please [use Reviewable](https://github.com/maxitg/SetReplace/blob/master/.github/CONTRIBUTING.md#code-review) in the future (link on the bottom of the PR message). It makes it easier to keep track of comments.

Reply:
That was mostly a statement of amusement, though as you notice a name like that does tend to suggest you stuffed to much meaning into a single variable. It happens.